// hapi-locals/lib/index.js
// Generated by CoffeeScript 1.9.3
(function() {
  var Hoek, context;

  Hoek = require("hoek");

  context = {};

  exports.register = function(server, options, next) {
    var locals;
    options = Hoek.applyToDefaults({
      isNullOverride: true,
      isMergeArrays: true
    }, options);
    locals = function(key, value) {
      return context[key] = value;
    };
    server.method({
      name: "locals",
      method: locals,
      options: {}
    });
    server.ext("onPostHandler", function(request, reply) {
      var destContext, response;
      response = request.response;
      destContext = {};
      if (response.variety === "view") {
        Hoek.merge(destContext, context, options.isNullOverride, options.isMergeArrays);
        Hoek.merge(destContext, response.source.context || {}, options.isNullOverride, options.isMergeArrays);
        response.source.context = destContext;
      }
      return reply["continue"]();
    });
    return next();
  };

  exports.register.attributes = {
    pkg: require("../package.json")
  };

}).call(this);
